Classes/Common/Helper.php

```diff
@@ -214,26 +236,58 @@ public static function devLog($message, $severity = 0)
         }
     }
 
+    /**
+     * Digest the given string
+     *
+     * @access public
+     *
+     * @param string $string: The string to encrypt
+     *
+     * @return mixed Hashed string or false on error
+     */
+    public static function digest($string)
+    {
+        if (!in_array(self::$hashAlgorithm, openssl_get_md_methods(true))) {
+            self::devLog('OpenSSL library doesn\'t support hash algorithm', DEVLOG_SEVERITY_ERROR);
+            return false;
+        }
+        // Hash string.
+        $hashed = openssl_digest($string, self::$hashAlgorithm);
+        return $hashed;
+    }
+
     /**
      * Encrypt the given string
      *
      * @access public
      *
      * @param string $string: The string to encrypt
      *
-     * @return array Array with encrypted string and control hash
+     * @return mixed Encrypted string or false on error
      */
     public static function encrypt($string)
     {
+        if (
+            !in_array(self::$cipherAlgorithm, openssl_get_cipher_methods(true))
+            || !in_array(self::$hashAlgorithm, openssl_get_md_methods(true))
+        ) {
+            self::devLog('OpenSSL library doesn\'t support cipher and/or hash algorithm', DEVLOG_SEVERITY_ERROR);
+            return false;
+        }
         if (empty($GLOBALS['TYPO3_CONF_VARS']['SYS']['encryptionKey'])) {
             self::devLog('No encryption key set in TYPO3 configuration', DEVLOG_SEVERITY_ERROR);
-            return;
+            return false;
+        }
+        // Generate random initialisation vector.
+        $iv = openssl_random_pseudo_bytes(openssl_cipher_iv_length(self::$cipherAlgorithm));
+        $key = openssl_digest($GLOBALS['TYPO3_CONF_VARS']['SYS']['encryptionKey'], self::$hashAlgorithm, true);
+        // Encrypt data.
+        $encrypted = openssl_encrypt($string, self::$cipherAlgorithm, $key, OPENSSL_RAW_DATA, $iv);
```

Comment:
This fails with the following exception (including binary data):

<pre>
PHP Warning: openssl_encrypt(): A tag should be provided when using AEAD mode in /var/www/extensions/dlf/Classes/Common/Helper.php line 285

at openssl_encrypt('dlfCore1', 'aes-256-gcm', '��NÒ�&���|��m���UuV�P�|�|uG(@', 1, 'Ie�:UsZ�')
</pre>

What am I missing? I've just have the search plugin on the page.

Reply:
Sorry, my fault! I experimented with different cipher algorithms and didn't push the final choice to upstream...

Explanation: The `*-gcm` and `*-ccm` cipher methods use authentication (AEAD mode) and therefore need a tag. But we don't need authentication, so I changed the cipher method to `AES-256-CTR`. This is the same as `AES-256-GCM`, but without authentication.

Reply:
Although this worked fine, I added base64-encoding to get rid of the binary data in HTML. I think it's less confusing now.